sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -94,14 +94,16 @@ class CacheManager extends Logging {
       logWarning("Asked to cache already cached data.")
     } else {
       val sparkSession = query.sparkSession
-      cachedData.add(CachedData(
-        planToCache,
-        InMemoryRelation(
-          sparkSession.sessionState.conf.useCompression,
-          sparkSession.sessionState.conf.columnBatchSize,
-          storageLevel,
-          sparkSession.sessionState.executePlan(planToCache).executedPlan,
-          tableName)))
+      val inMemoryRelation = InMemoryRelation(
+        sparkSession.sessionState.conf.useCompression,
+        sparkSession.sessionState.conf.columnBatchSize,
+        storageLevel,
+        sparkSession.sessionState.executePlan(planToCache).executedPlan,
+        tableName)
+      if (planToCache.conf.cboEnabled && planToCache.stats.rowCount.isDefined) {
```

Comment:
Do we need to limit to those conditions? I think we can pass the stats into the created `InMemoryRelation` even the two conditions don't match.

Reply:
the reason I put it here is that when we did not enable CBO, the stats in the underlying plan might be much smaller than the actual size in memory leading to the potential risk of OOM error. 

The underlying cause is that without CBO enabled, the size of the plan is calculated with BaseRelation's sizeInBytes, but with CBO, we can have a more accurate estimation, 


https://github.com/apache/spark/blob/03fdc92e42d260a2b7c0090115f162ba5c091aae/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L42-L46

https://github.com/apache/spark/blob/03fdc92e42d260a2b7c0090115f162ba5c091aae/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala#L370-L381

Reply:
When CBO is disabled, don't we just set the `sizeInBytes` to `defaultSizeInBytes`? Isn't it different than current statistics of first run?

Reply:
no, if CBO is disabled, the relation's sizeInBytes is the file size

https://github.com/apache/spark/blob/5c3a1f3fad695317c2fff1243cdb9b3ceb25c317/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala#L85

Reply:
`LogicalRelation` uses the statistics from the `relation` only when there is no given `catalogTable`.  In this case, it doesn't consider if CBO is enabled or not.

Only `catalogTable` considers CBO when computing its statistics in `toPlanStats`. It doesn't refer to relation's statistics, IIUC.

Reply:
If a catalog table doesn't have statistics in its metadata, we will fill it with `defaultSizeInBytes`.

https://github.com/apache/spark/blob/326f1d6728a7734c228d8bfaa69442a1c7b92e9b/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala#L121-L134

Reply:
@viirya  you're right! Thanks for clearing the confusion

however, to prevent using relation's stats which can be much smaller than the in-memory size and lead to a potential OOM error, we should still have this condition here (we can remove cboEnabled though), right?

Reply:
The statistics from relation is based on files size, will it easily cause OOM issue? I think in the cases other than cached query, we still use this relation's statistics. If this is an issue, doesn't it also affect the other cases?

Reply:
it's true, it affects I believe....there is a similar discussion in https://github.com/apache/spark/pull/19743